modules/apps/remote-web-app/remote-web-app-web/src/main/java/com/liferay/remote/web/app/web/internal/RemoteWebAppPortletRegistrar.java

```diff
@@ -0,0 +1,136 @@
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.remote.web.app.web.internal;
+
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.remote.web.app.model.RemoteWebAppEntry;
+import com.liferay.remote.web.app.service.RemoteWebAppEntryLocalService;
+import com.liferay.remote.web.app.web.internal.portlet.RemoteWebAppPortlet;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+@Component(immediate = true, service = RemoteWebAppPortletRegistrar.class)
+public class RemoteWebAppPortletRegistrar {
+
+	public void registerPortlet(RemoteWebAppEntry remoteWebAppEntry) {
+		_registerPortlet(
+			remoteWebAppEntry.getEntryId(), remoteWebAppEntry.getName(),
+			remoteWebAppEntry.getUrl());
+	}
+
+	public void unregisterPortlet(RemoteWebAppEntry remoteWebAppEntry) {
+		_unregisterPortlet(remoteWebAppEntry.getEntryId());
+	}
+
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_bundleContext = bundleContext;
+
+		if (_log.isInfoEnabled()) {
+			_log.info("Starting remote web apps");
+		}
+
+		_registerPortlet(0, "El País", "http://www.elpais.com");
+
+		/*
+
+		for (RemoteWebAppEntry remoteWebAppEntry :
+				remoteWebAppEntryLocalService.getRemoteWebAppEntries(
+					0, QueryUtil.ALL_POS)) {
+
+			registerPortlet(remoteWebAppEntry);
+		}
+		*/
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		if (_log.isInfoEnabled()) {
+			_log.info("Stopping all remote web apps");
+		}
+
+		Collection<RemoteWebAppPortlet> remoteWebAppPortlets;
+
+		synchronized (_remoteWebAppPortlets) {
+			remoteWebAppPortlets = _remoteWebAppPortlets.values();
+
+			_remoteWebAppPortlets.clear();
+		}
+
+		for (RemoteWebAppPortlet remoteWebAppPortlet : remoteWebAppPortlets) {
+			remoteWebAppPortlet.unregister();
+		}
+	}
+
+	@Reference
+	protected RemoteWebAppEntryLocalService remoteWebAppEntryLocalService;
+
+	private void _registerPortlet(long id, String name, String url) {
+		RemoteWebAppPortlet remoteWebAppPortlet = new RemoteWebAppPortlet(
+			id, name, url);
+
+		synchronized (_remoteWebAppPortlets) {
+			if (_remoteWebAppPortlets.containsKey(id)) {
+				throw new IllegalStateException(
+					"Portlet " + id + " is already registered");
+			}
+
+			_remoteWebAppPortlets.put(id, remoteWebAppPortlet);
+
+			remoteWebAppPortlet.register(_bundleContext);
```

Comment:
The important logic has been moved to the portlet class, to centralize it there.